pkg/storage/db.go

```diff
@@ -473,8 +475,125 @@ func (db *DB) Store(ctx context.Context, cloudAssetChanges domain.CloudAssetChan
 		}
 		return err
 	}
-	return tx.Commit()
+	if err = tx.Commit(); err != nil {
+		return err
+	}
+	ver, err := db.GetSchemaVersion(ctx)
+	if err != nil || ver < DualWriteSchemaVersion {
+		return nil
+	}
+	return db.StoreV2(ctx, cloudAssetChanges)
+}
 
+// StoreV2 Currently public to allow testing separately.Storage interface implementation that records to a database using new schema, to replace Store in the future.
+func (db *DB) StoreV2(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	tx, err := db.sqldb.Begin()
+	if err != nil {
+		return err
+	}
+	if err = db.ensureResourceExists(ctx, cloudAssetChanges, tx); err == nil {
+		arnID := resIDFromARN(cloudAssetChanges.ARN)
+		for _, val := range cloudAssetChanges.Changes {
+			for _, ip := range val.PrivateIPAddresses {
+				if strings.EqualFold(added, val.ChangeType) {
+					err = db.assignPrivateIP(ctx, tx, arnID, ip, cloudAssetChanges.ChangeTime)
+				} else {
+					err = db.releasePrivateIP(ctx, tx, arnID, ip, cloudAssetChanges.ChangeTime)
+				}
+				if err != nil {
+					break
+				}
+			}
+			for ix, ip := range val.PublicIPAddresses {
+				hostname := val.Hostnames[ix]
+				if strings.EqualFold(added, val.ChangeType) {
+					err = db.assignPublicIP(ctx, tx, arnID, ip, hostname, cloudAssetChanges.ChangeTime)
+				} else {
+					err = db.releasePublicIP(ctx, tx, arnID, ip, hostname, cloudAssetChanges.ChangeTime)
+				}
+			}
+			if err != nil {
+				break
+			}
+		}
+	}
+	if err != nil {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			return errors.Wrap(rollbackErr, err.Error()) // so we don't lose the original error
+		}
+		return err
+	}
+	if err = tx.Commit(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (db *DB) ensureResourceExists(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges, tx *sql.Tx) error {
+	// reading pre-requisite to understand this query - https://www.postgresql.org/docs/current/queries-with.html
+	const createResourceQuery string = `
+with sel as (
```

Comment:
nit: notably other SQL in the codebase uses a mix of CAPS and lowercase for queries e.g. https://github.com/asecurityteam/asset-inventory-api/pull/38/files#diff-106eb1177cdcbe545695a4fdec2e2a32L48 I don't think we have a SQL style guideline anywhere that dictates we can't do lowercase so I'm ok with lowercase sql but the lack of uniformity will be confusing when onboarding someone. Eventually I'd like us to choose all lowercase or mixed casing.

Reply:
We should be left with "new" SQL only by the time we are done with  remaining schema change related work. 
I am open to both lower-case and uppercase too though.